Add GSASII tab to EngDiff interface

DanielMurphy22 opened the pull request:
Note this includes a quick fix for issue #34262, but worth checking that issue isn't more complex.

**To test:**

You may wish to test this on a development IDAaaS environment as this will have gsas2 installed.

Data to test is in babylon>Public>Daniel Murphy>GSASII tab. Note the CIF file is from the Mantid code base and the instrument and focused data for the tests below can be generated by running the Calibration and Focus tabs (and they will prefill the GSASII tab):

```
# Calibration
/cycle_19_1/ENGINX00305738.nxs
# Sample Run
/cycle_19_1/ENGINX00305761.nxs
# Vanadium
/cycle_19_1/ENGINX00307521.nxs
```

Test 1: Bank 1
```
 # Instrument Group
ENGINX_305738_bank_1.prm
# Phase
FE_GAMMA.cif
# Focused Data
ENGINX_305761_307521_bank_1_TOF.gss
```

Test 2: All banks
```
 # Instrument Group
ENGINX_305738_all_banks.prm
# Phase
FE_GAMMA.cif
# Focused Data
ENGINX_305761_307521_all_banks_TOF.gss
```

Test 3: All banks, separate Focused files
```
 # Instrument Group
ENGINX_305738_all_banks.prm
# Phase
FE_GAMMA.cif
# Focused Data
ENGINX_305761_307521_bank_1_TOF.gss, ENGINX_305761_307521_bank_2_TOF.gss
```

Test 4: All banks, separate Focused and Instrument files
```
 # Instrument Group
ENGINX_305738_bank_1.prm, ENGINX_305738_bank_2.prm
# Phase
FE_GAMMA.cif
# Focused Data
ENGINX_305761_307521_bank_1_TOF.gss, ENGINX_305761_307521_bank_2_TOF.gss
```

(I currently don't want to allow one focused data file with many instrument files)

I have added reitveld data from the GSASII tutorial on babylon but running those files will not work as my assumptions have focused on the output file types from the EngDiff interface.

- When you have run a refinement you can change the fitting range by dragging the start and end X lines or by altering the min and max line edits. If you refine again with the same load parameters then the new fitting range will be used. In any of the all bank examples you will refine 2 spectra, but only the first is plotted, to plot the second toggle the combobox just above the plot. For me, the best fit for the examples above was setting the override cell length to 3.65.


Fixes #34124

Not yet added ~~release notes~~ or docs!!



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Add GSASII tab to EngDiff interface

Note this includes a quick fix for issue 34262

DanielMurphy22 committed:
Fix Focus and Settings tests

commented out gsas2 test files
minor code tidy

DanielMurphy22 commented:
I have tried to build a linux package from this branch which would aide testing on IDAaaS: https://builds.mantidproject.org/job/build_packages_from_branch/77/

DanielMurphy22 committed:
Code tidy model

Stop mantid_pawley_refelctions being referenced before assigned
Remove CtestCostData
Comment out testing defaults
do not raise errors log them and return back to the presenter
catch refinement if an input filepath is invalid
set default path_to_gsas2 to work on IDAaaS
use softcoded path to call_G2sc.py
pass string to Popen to work on Windows
Add invalid * to Project name input
Add invalid * warning to checkboxes

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/presenter.py:
```diff
@@ -0,0 +1,91 @@
+    def get_limits_if_same_load_parameters(self):
+        new_load_params = self.view.get_load_parameters()
+        if not new_load_params:
+            return None
```
I tried hitting the "Refine in GSAS II" button with an invalid Focused Data file and I got an error "Could not determine Minimum and Maximum X values ...". I wonder if this check here should have caught this? The `new_load_params` variable was a list containing 3 items and the 3rd one was empty

DanielMurphy22 replied:
I have caught this in the model instead by checking that each of the filepath parameters are not None

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/presenter.py:
```diff
@@ -0,0 +1,91 @@
+    def save_latest_load_parameters(self):
+        self.latest_load_parameters = self.view.get_load_parameters()
+
+    def get_limits_if_same_load_parameters(self):
```
What's the reason for only using the user supplied limits if the load parameters are the same? Seems like if it's the first load and the user has specified an x min and max then they should still be used?

DanielMurphy22 replied:
I think it's hard to know what it should do. I had settled on the idea that a new refinement would use the full range (as I determine from the input files) and the user could then change it and re-run the same refinement. If we changed to allow the user to set the limits before the first refinement, then we need a way for them to "clear" the limits before refining something new. Maybe we should only check if the instrument file has changed, as a new focused data file on the same instrument may need a similar range. This would still need a clearing. We can discuss!

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/settings/settings_presenter.py:
```diff
@@ -20,7 +21,10 @@
     "sort_ascending": True,
-    "default_peak": "BackToBackExponential"
+    "default_peak": "BackToBackExponential",
+    "path_to_gsas2": "/home/danielmurphy/gsas2",
```
I think this default should be removed

DanielMurphy22 replied:
Changed to /opt/gsas2/ as on IDAaaS

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/call_G2sc.py:
```diff
@@ -0,0 +1,211 @@
+'''Call GSASIIscriptable'''
+
+try:
+    import_path = path_to_gsas2 + 'GSASII'
```
should this use os.path.join like you've done a few lines lower down?

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py:
```diff
@@ -0,0 +1,572 @@
+        )
+
+        call_gsas2 = (path_to_gsas2 + "bin/python "
+                      + "/home/danielmurphy/mantid/qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/"
```
The GSAS call currently fails for me because you've got this path which is specific to your machine. Can this be made a relative path?
Also I think the logic to append `path_to_gsas2` and bin\python could benefit from using os.path.join in case the user doesn't have a closing slash on the path in settings
Finally in my local GSAS II installation I don't have a bin directory. The python.exe is just sitting in the root directory of the install which in my case is `C:\Users\iao46725\gsas2full`. This is GSAS II v5203. Maybe that's not the right version...

DanielMurphy22 replied:
Fixed this and removed the used simply the gsas2_path + python for Windows

DannyHindson reviewed (changes requested):
I've had an initial look around. The problems in model.py meant I didn't get the call to GSAS II to work. I was trying to test on Windows with a local GSAS II install. Maybe I'll try on IDAaaS and see if that works

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py:
```diff
@@ -0,0 +1,572 @@
+
+    def call_subprocess(self, command_string):
+        try:
+            shell_process = subprocess.Popen([command_string.replace('"', '\\"')],
```
For me on Windows the presence of the `[` and `]` characters around the command was causing the call to GSAS to fail with this error:
`The filename, directory name, or volume label syntax is incorrect.`
If I remove the square brackets this problem goes away. Looking at the documentation for the python subprocess class (https://docs.python.org/3/library/subprocess.html) I think this is perhaps a Windows thing because there's some Windows specific handling when the args parameters is a list. Having said that perhaps the command_string variable isn't a comma separated list that would become a python list when wrapping in square brackets. So maybe the square brackets could be dropped without any downside on all platforms

DanielMurphy22 replied:
I thought I'd seen advice somewhere that said to use a list with Popen, but it does work without!

DanielMurphy22 committed:
Add gsas2_presenter tests

DanielMurphy22 committed:
Add GSAS2 Model tests

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py:
```diff
@@ -0,0 +1,701 @@
+                                     str(overriding_cell_lengths_list[1]),
+                                     str(overriding_cell_lengths_list[2])])
+        else:
+            cell_length_a = self.find_in_file(phase_file_path, '_cell_length_a', ' ', '_cell_length_b')
```
If an invalid phase file is supplied these calls to `find_in_file` will return None so please can you check for None and return an "invalid phase file format" error or similar? The join command on L411 falls over if any of the cell_length_x variables are None

DannyHindson reviewed (changes requested):
Just a couple of things so far (one of which is the concatenation of the two strings in the error message when the GSAS call fails).
Also when I run a refinement I get this message written to the log. No big deal but do you know what is causing this?
![image](https://user-images.githubusercontent.com/56295817/186709620-02a68bfd-7b42-473a-a48d-5fbbeb03e260.png)

DannyHindson commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py:
```diff
@@ -0,0 +1,701 @@
+                crystal_params.append(loop_crystal_params)
+        tof_min = []
+        for loop_crystal_param_string in crystal_params:
+            list_crystal_params = loop_crystal_param_string.split("\t")
```
If I run a refinement with instrument group=bank 2, focused data=all banks I get an error from this line (`AttributeError: 'NoneType' object has no attribute 'split'`). I think one of the earlier calls to `find_in_file` on L371 or 375 returns None and this gets entered into the crystal_params list when there's a mismatch in the number of banks in the prm and gss files so worth checking if find_in_file returns None here

DanielMurphy22 committed:
Add test_files

DanielMurphy22 committed:
Requested Changes

Protect against invalid phase file
Protect against mismatch number of banks in data and inst files
Add Pythonpath to fix subprocess call on Windows

DanielMurphy22 commented:
I would note that I also had on my list to add tests for call_G2sc.add_histogram() and EnggUtils._save_output_files() (to check that SaveGSS has append=False). These are hopefully not needed for this PR to be merged before code freeze.

DannyHindson committed:
Use FileFinder class to simplify search paths

This removes assumptions about content of the datasearch.directories
setting in the .properties file
Also remove dependency on a couple of system test ref files that
won't be present when running unit tests on Jenkins

DannyHindson committed:
Add release note

DannyHindson commented:
I've reviewed this PR and have tested on Windows and Ubuntu and it works fine. There are possibly a couple of outstanding pieces of work that can be picked up on separate PRs and I'll discuss with Dan when he's back from holiday:
- testing on IDAaaS. I haven't done this as part of my review but would like to check if Dan's done any of this
- documentation. It's possible that it makes sense to do this once the UI has had some further use by Saurabh and the ENGINX team to avoid having to update screenshots multiple times

RichardWaiteSTFC commented:
Just tried testing this on IDAaaS (using .tar.xz found here https://builds.mantidproject.org/job/build_packages_from_branch/85/)
I got the following error when trying to refine the bank 1 data
```
Line not defined:     52832.78540        0.00000000        0.00000000
Line not defined:     END
Error in execution of algorithm LoadGSS:
All arguments to init must be positive and non-zero
Traceback (most recent call last):
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/presenter.py", line 36, in on_refine_clicked
    number_output_histograms = self.model.run_model(load_params, refine_params, project_name,
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py", line 94, in run_model
    self.evaluate_further_inputs(user_limits)
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py", line 138, in evaluate_further_inputs
    if not self.loop_phase_files() or not self.validate_x_limits(user_x_limits):
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py", line 512, in validate_x_limits
    self.understand_data_structure()
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/gsas2/model.py", line 503, in understand_data_structure
    loop_focused_workspace = LoadGSS(Filename=input_file, OutputWorkspace="GSASII_input_data")
  File "/home/rw1061637/Downloads/mantidworkbenchunstable/lib/python3.8/site-packages/mantid/simpleapi.py", line 1058, in __call__
    algm.execute()
IndexError: All arguments to init must be positive and non-zero
```
To Reproduce (may not need all these steps but this is what I did):
(1) Load .prm in calibration tab (auto-filled in GSAS tab nicely!) - this prodsuced the following error
```
Unable to load calibration file /babylon/Public/DanielMurphy/GSASII Tab/ENGINX_305738_bank_1.nxs. Error: Problem setting "Filename" in Load: Invalid value for property Filename (list of str lists) from string "/babylon/Public/DanielMurphy/GSASII Tab/ENGINX_305738_bank_1.nxs": When setting value of property "Slave": File "/babylon/Public/DanielMurphy/GSASII Tab/ENGINX_305738_bank_1.nxs" not found
If the file has been found but you got this error, you might not have read permissions or the file might be corrupted.
If the file has not been found, you might have forgotten to add its location in the data search directories.
```
Even though I pointed it at the `.prm`.
It loaded the following workspaces (which I think look correct)
![image](https://user-images.githubusercontent.com/55979119/188411535-747138d3-3aad-4239-bab0-04a1b1c90ca2.png)

(2) Switch to GSAS II tab
(3) fill in project name
(4) Set paths the phase and data
(5) Change refinement setting as so
![image](https://user-images.githubusercontent.com/55979119/188410437-5e05b37b-7468-4c99-88d5-89bacced2414.png)
(6) Click Refine in GSAS II

Did I do something wrong?

I a

DannyHindson commented:
I've done some testing on IDAaaS with a new build that contains all of the fixes up to and including 1st Sept:
https://builds.mantidproject.org/job/build_packages_from_branch/93/

There seems to be a problem relating to the subprocess.POpen command. On IDAaaS it removes the curly brackets at the start and end of the JSON string that is included in the command_string variable. I've just checked on Ubuntu and the latest code still works on there so I think this is a RHEL only problem. There are some comments online about POpen doing this although not specifically on RHEL7 - the comments suggest that the "{" and "}" commands have a particular meaning in certain shells and get removed.

I was able to make some changes to the .py files installed to debug the problem and to try out a few alternatives. It seems to be the case that putting single quotes around the JSON string stops the curly brackets being removed. I also had to remove the code in gsas2/model.py to replace `"` with `\\"`. Perhaps once the whole JSON string is inside single quotes the double quotes that bound the property names don't need escaping any more. The POpen call continues to work on Ubuntu with this change so will push a commit shortly.

DannyHindson commented:
> 
> Did I do something wrong?

I think the first error about not being able to find a .nxs file is because the file doesn't exist on Dan's babylon share. The calibration tab is looking for a table workspace containing the calibration parameters (difc etc). I think he was probably expecting us to go straight to the GSAS2 tab and load the .prm file there so he didn't upload all of the calibration outputs onto the share. 
As discussed the LoadGSS problem is perhaps because the format of the bank 1 .gss file on Dan's share has a mistake in it. The BANK line has a # character at the start:
```
# BANK 1 10186 10186 RALF   171834       39   171834 0.00023 FXYE
```

DannyHindson committed:
Stop using shell in subprocess Popen call

Was causing problems due to different behaviour on different platforms,
especially in relation to the JSON string holding the refinement parameters
eg curly braces removed on RHEL7, Windows didn't like surrounding the
JSON string in single quotes
Supply a list instead of a string and this allows Shell=False to be used

DannyHindson reviewed (approved):
Approving following recent successful test on IDAaaS. I did a couple of tidy ups on this at the end given that Dan is away on holiday but not large enough to necessitate someone else reviewing\approving
